src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1740,10 +1741,20 @@ private void CheckContainsDocument(DocumentId documentId)
             throw new ArgumentNullException(nameof(documentId));
         }
 
-        if (!ContainsDocument(documentId))
+        // For source generated documents we expect them to be already generated to use any of the APIs that call this
+        if (documentId.IsSourceGenerated &&
+            ContainsProject(documentId.ProjectId) &&
+            this.GetRequiredProject(documentId.ProjectId).TryGetSourceGeneratedDocumentForAlreadyGeneratedId(documentId) is not null)
         {
-            throw new InvalidOperationException(WorkspaceExtensionsResources.The_solution_does_not_contain_the_specified_document);
+            return;
         }
```

Comment:
i think i'm still confused though.  you have `is not null`.  and you're doing `TryGetSourceGeneratedDocumentForAlreadyGeneratedId`.  I guess that means it can return `null, true, false`?  But i don't quite get why this is fine to get `false` back.  Might be worth clarifying.

Reply:
It returns null if the document isn't found. I've extracted this to a local function to make it clearer